### PR TITLE
[autoparallel] move ckpt solvers to autoparallel folder / refactor code

### PR DESCRIPTION
**Pull request by super-dainiu**
### What's new?
I provided a new base class for activation checkpoint solvers. Also, I moved the activation checkpoint annotations into the `node.meta`, because this is more original. In this way, we can `deepcopy` the `Graph` without losing information.
To prepare data for our solver, we must do `MetaInfoProp` first.
```python
model = tm.resnet18()
tracer = ColoTracer()
graph = tracer.trace(model)
gm = ColoGraphModule(model, graph)
interp = MetaInfoProp(gm.cuda())
interp.propagate(MetaTensor(torch.rand(128, 3, 224, 224, device='meta'), fake_device='cuda:0'))
```
The first solver is `CheckpointSolverChen(graph: Graph, cnode: List[str] = None, num_grids: int = 6)`. It can be used with the following code.

```python
solver = CheckpointSolverChen(gm.graph)
chen_graph = solver.solve()
```
The second solver is `CheckpointSolverRotor(graph: Graph, memory_budget: float = -1, parameter_size: float = 0, cnode: List[str] = None, memory_slots: int = 500)`. It can be used with the following code.
```python
solver = CheckpointSolverRotor(gm.graph, memory_budget=2100 * 1024**2, parameter_size=parameter_size(gm))
rotor_graph = solver.solve(force_python=True) # I haven't refactored the C code yet.
```

Then assign the generated graph to the graph module, which is optimized with its code.

```python
gm.graph = rotor_graph
print(gm.code)
```
With our customized `ActivationCheckpointCodeGen()`
```python
def checkpoint_0(self, x):
    conv1 = self.conv1(x);  x = None
    bn1 = self.bn1(conv1);  conv1 = None
    relu = self.relu(bn1);  bn1 = None
    maxpool = self.maxpool(relu);  relu = None
    return maxpool

def forward(self, x : torch.Tensor) -> torch.Tensor:
    maxpool = colossalai.utils.activation_checkpoint.checkpoint(self.checkpoint_0, False, x, use_reentrant=False)
    layer1_0_conv1 = getattr(self.layer1, "0").conv1(maxpool)
    layer1_0_bn1 = getattr(self.layer1, "0").bn1(layer1_0_conv1);  layer1_0_conv1 = None
    layer1_0_relu = getattr(self.layer1, "0").relu(layer1_0_bn1);  layer1_0_bn1 = None
    layer1_0_conv2 = getattr(self.layer1, "0").conv2(layer1_0_relu);  layer1_0_relu = None
    layer1_0_bn2 = getattr(self.layer1, "0").bn2(layer1_0_conv2);  layer1_0_conv2 = None
    add = layer1_0_bn2 + maxpool;  layer1_0_bn2 = maxpool = None
    layer1_0_relu_1 = getattr(self.layer1, "0").relu(add);  add = None
    layer1_1_conv1 = getattr(self.layer1, "1").conv1(layer1_0_relu_1)
    layer1_1_bn1 = getattr(self.layer1, "1").bn1(layer1_1_conv1);  layer1_1_conv1 = None
    layer1_1_relu = getattr(self.layer1, "1").relu(layer1_1_bn1);  layer1_1_bn1 = None
    layer1_1_conv2 = getattr(self.layer1, "1").conv2(layer1_1_relu);  layer1_1_relu = None
    layer1_1_bn2 = getattr(self.layer1, "1").bn2(layer1_1_conv2);  layer1_1_conv2 = None
    add_1 = layer1_1_bn2 + layer1_0_relu_1;  layer1_1_bn2 = layer1_0_relu_1 = None
    layer1_1_relu_1 = getattr(self.layer1, "1").relu(add_1);  add_1 = None
    layer2_0_conv1 = getattr(self.layer2, "0").conv1(layer1_1_relu_1)
    layer2_0_bn1 = getattr(self.layer2, "0").bn1(layer2_0_conv1);  layer2_0_conv1 = None
    layer2_0_relu = getattr(self.layer2, "0").relu(layer2_0_bn1);  layer2_0_bn1 = None
    layer2_0_conv2 = getattr(self.layer2, "0").conv2(layer2_0_relu);  layer2_0_relu = None
    layer2_0_bn2 = getattr(self.layer2, "0").bn2(layer2_0_conv2);  layer2_0_conv2 = None
    layer2_0_downsample_0 = getattr(getattr(self.layer2, "0").downsample, "0")(layer1_1_relu_1);  layer1_1_relu_1 = None
    layer2_0_downsample_1 = getattr(getattr(self.layer2, "0").downsample, "1")(layer2_0_downsample_0);  layer2_0_downsample_0 = None
    add_2 = layer2_0_bn2 + layer2_0_downsample_1;  layer2_0_bn2 = layer2_0_downsample_1 = None
    layer2_0_relu_1 = getattr(self.layer2, "0").relu(add_2);  add_2 = None
    layer2_1_conv1 = getattr(self.layer2, "1").conv1(layer2_0_relu_1)
    layer2_1_bn1 = getattr(self.layer2, "1").bn1(layer2_1_conv1);  layer2_1_conv1 = None
    layer2_1_relu = getattr(self.layer2, "1").relu(layer2_1_bn1);  layer2_1_bn1 = None
    layer2_1_conv2 = getattr(self.layer2, "1").conv2(layer2_1_relu);  layer2_1_relu = None
    layer2_1_bn2 = getattr(self.layer2, "1").bn2(layer2_1_conv2);  layer2_1_conv2 = None
    add_3 = layer2_1_bn2 + layer2_0_relu_1;  layer2_1_bn2 = layer2_0_relu_1 = None
    layer2_1_relu_1 = getattr(self.layer2, "1").relu(add_3);  add_3 = None
    layer3_0_conv1 = getattr(self.layer3, "0").conv1(layer2_1_relu_1)
    layer3_0_bn1 = getattr(self.layer3, "0").bn1(layer3_0_conv1);  layer3_0_conv1 = None
    layer3_0_relu = getattr(self.layer3, "0").relu(layer3_0_bn1);  layer3_0_bn1 = None
    layer3_0_conv2 = getattr(self.layer3, "0").conv2(layer3_0_relu);  layer3_0_relu = None
    layer3_0_bn2 = getattr(self.layer3, "0").bn2(layer3_0_conv2);  layer3_0_conv2 = None
    layer3_0_downsample_0 = getattr(getattr(self.layer3, "0").downsample, "0")(layer2_1_relu_1);  layer2_1_relu_1 = None
    layer3_0_downsample_1 = getattr(getattr(self.layer3, "0").downsample, "1")(layer3_0_downsample_0);  layer3_0_downsample_0 = None
    add_4 = layer3_0_bn2 + layer3_0_downsample_1;  layer3_0_bn2 = layer3_0_downsample_1 = None
    layer3_0_relu_1 = getattr(self.layer3, "0").relu(add_4);  add_4 = None
    layer3_1_conv1 = getattr(self.layer3, "1").conv1(layer3_0_relu_1)
    layer3_1_bn1 = getattr(self.layer3, "1").bn1(layer3_1_conv1);  layer3_1_conv1 = None
    layer3_1_relu = getattr(self.layer3, "1").relu(layer3_1_bn1);  layer3_1_bn1 = None
    layer3_1_conv2 = getattr(self.layer3, "1").conv2(layer3_1_relu);  layer3_1_relu = None
    layer3_1_bn2 = getattr(self.layer3, "1").bn2(layer3_1_conv2);  layer3_1_conv2 = None
    add_5 = layer3_1_bn2 + layer3_0_relu_1;  layer3_1_bn2 = layer3_0_relu_1 = None
    layer3_1_relu_1 = getattr(self.layer3, "1").relu(add_5);  add_5 = None
    layer4_0_conv1 = getattr(self.layer4, "0").conv1(layer3_1_relu_1)
    layer4_0_bn1 = getattr(self.layer4, "0").bn1(layer4_0_conv1);  layer4_0_conv1 = None
    layer4_0_relu = getattr(self.layer4, "0").relu(layer4_0_bn1);  layer4_0_bn1 = None
    layer4_0_conv2 = getattr(self.layer4, "0").conv2(layer4_0_relu);  layer4_0_relu = None
    layer4_0_bn2 = getattr(self.layer4, "0").bn2(layer4_0_conv2);  layer4_0_conv2 = None
    layer4_0_downsample_0 = getattr(getattr(self.layer4, "0").downsample, "0")(layer3_1_relu_1);  layer3_1_relu_1 = None
    layer4_0_downsample_1 = getattr(getattr(self.layer4, "0").downsample, "1")(layer4_0_downsample_0);  layer4_0_downsample_0 = None
    add_6 = layer4_0_bn2 + layer4_0_downsample_1;  layer4_0_bn2 = layer4_0_downsample_1 = None
    layer4_0_relu_1 = getattr(self.layer4, "0").relu(add_6);  add_6 = None
    layer4_1_conv1 = getattr(self.layer4, "1").conv1(layer4_0_relu_1)
    layer4_1_bn1 = getattr(self.layer4, "1").bn1(layer4_1_conv1);  layer4_1_conv1 = None
    layer4_1_relu = getattr(self.layer4, "1").relu(layer4_1_bn1);  layer4_1_bn1 = None
    layer4_1_conv2 = getattr(self.layer4, "1").conv2(layer4_1_relu);  layer4_1_relu = None
    layer4_1_bn2 = getattr(self.layer4, "1").bn2(layer4_1_conv2);  layer4_1_conv2 = None
    add_7 = layer4_1_bn2 + layer4_0_relu_1;  layer4_1_bn2 = layer4_0_relu_1 = None
    layer4_1_relu_1 = getattr(self.layer4, "1").relu(add_7);  add_7 = None
    avgpool = self.avgpool(layer4_1_relu_1);  layer4_1_relu_1 = None
    flatten = torch.flatten(avgpool, 1);  avgpool = None
    fc = self.fc(flatten);  flatten = None
    return fc
```
### Tests
image

### CODE REVIEW

File: colossalai/auto_parallel/checkpoint/ckpt_solver_base.py [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+from colossalai.fx.profiler.memory_utils import is_inplace
+
+__all___ = ['CheckpointSolverBase']
+
+
+def _copy_output(src: Graph, dst: Graph):
+    """Copy the output node from src to dst"""
+    for n_src, n_dst in zip(src.nodes, dst.nodes):
+        if n_src.op == 'output':
+            n_dst.meta = n_src.meta
```
Comment (super-dainiu) [inline colossalai/auto_parallel/checkpoint/ckpt_solver_base.py:17]: A bad behavior of `deepcopy(graph)` is that the graph output is changed.

Comment (super-dainiu) [reply colossalai/auto_parallel/checkpoint/ckpt_solver_base.py:17]: So we need to put all the meta information back to the copied nodes.

File: colossalai/auto_parallel/checkpoint/ckpt_solver_rotor.py [hunk 2]
```diff
@@ -0,0 +1,378 @@
...
+                        else:
+                            cost_table[m][i][idx] = chain_checkpoint
+                            back_ptr[m][i][idx] = (True,)
+        return cost_table, back_ptr
+
+    @staticmethod
+    def _compute_table_c(chain: Chain, mem_slots: int) -> Tuple:
+        raise NotImplementedError("C implementation not available yet")
+
+    def _backtrack(self, chain: Chain, lmin: int, lmax: int, mem_budget: int, cost_table: List[List[Dict[int, Tuple]]],
```
Comment (super-dainiu) [inline colossalai/auto_parallel/checkpoint/ckpt_solver_rotor.py:245]: Can check these functions and namings. cc @Cypher30

File: colossalai/auto_parallel/checkpoint/ckpt_solver_rotor.py [hunk 3]
```diff
@@ -0,0 +1,378 @@
...
+            j = back_ptr[mem_budget][lmin][lmax][1]
+            sequence.insert(ForwardCheck(lmin))
+            for k in range(lmin + 1, j):
+                sequence.insert(ForwardNograd(k))
+            sequence.insert_sequence(self._backtrack(chain, j, lmax, mem_budget - chain.xbar[j], cost_table, back_ptr))
+            sequence.insert_sequence(self._backtrack(chain, lmin, j - 1, mem_budget, cost_table, back_ptr))
+        return sequence
+
+    @staticmethod
+    def _annotate_from_sequence(sequence: Sequence, node_list: List[List[Node]]):
```
Comment (super-dainiu) [inline colossalai/auto_parallel/checkpoint/ckpt_solver_rotor.py:292]: is this function concise?

File: colossalai/fx/codegen/activation_checkpoint_codegen.py [hunk 4]
```diff
@@ -1,14 +1,37 @@
-import colossalai
+from typing import Any, Callable, Dict, Iterable, List, Tuple
```
Comment (super-dainiu) [inline colossalai/fx/codegen/activation_checkpoint_codegen.py:1]: This file I just modify the annotation of checkpoints.
`node.activation_checkpoint` into `node.meta['activation_checkpoint']`

File: tests/test_fx/test_ckpt_solvers/test_ckpt_torchvision.py [hunk 5]
```diff
@@ -94,6 +95,7 @@ def _run_ckpt_solver(rank):
     gpc.destroy()
 
 
+@pytest.mark.skip("TODO(super-dainiu): refactor all tests.")
```
Comment (super-dainiu) [inline tests/test_fx/test_ckpt_solvers/test_ckpt_torchvision.py:98]: I will refactor tests next

File: colossalai/auto_parallel/checkpoint/ckpt_solver_rotor.py (hunk 2 above)
Comment (Cypher30) [reply colossalai/auto_parallel/checkpoint/ckpt_solver_rotor.py:254]: I think those namings are ok

File: colossalai/auto_parallel/checkpoint/ckpt_solver_rotor.py (hunk 3 above)
Comment (Cypher30) [reply colossalai/auto_parallel/checkpoint/ckpt_solver_rotor.py:301]: This function might be more concise, I think this part could be left for me to refactor as I might need to refactor the activation codegen

Comment (super-dainiu) [reply colossalai/auto_parallel/checkpoint/ckpt_solver_rotor.py:301]: Our codegen is too heavy, we might dive deep later.

File: colossalai/fx/profiler/tensor.py [hunk 6]
```diff
@@ -12,10 +12,11 @@
...
 
 
-def set_uuid(x):
+def set_data_ptr(x):
     if isinstance(x, torch.Tensor):
-        if not hasattr(x, 'uuid'):
-            setattr(x, 'uuid', uuid.uuid4())
+        if not x.data_ptr():
+            data_ptr = uuid.uuid4()
+            x.data_ptr = lambda: data_ptr
```
Comment (super-dainiu) [inline colossalai/fx/profiler/tensor.py:19]: I also changed `tensor.uuid` to `tensor.data_ptr()`

**Review by FrankLeeeee — APPROVED** [review]

**Pull request merged**